src/reference/asciidoc/kafka.adoc

```diff
@@ -2950,6 +2945,46 @@ public static class Config {
 ----
 ====
 
+[[exactly-once]]
+==== Exactly Once Semantics
+
+You can provide a listener container with a `KafkaAwareTransactionManager` instance.
+When so configured, the container starts a transaction before invoking the listener.
+Any `KafkaTemplate` operations performed by the listener participate in the transaction.
+If the listener successfully processes the record (or multiple records, when using a `BatchMessageListener`), the container sends the offset(s) to the transaction by using `producer.sendOffsetsToTransaction()`), before the transaction manager commits the transaction.
+If the listener throws an exception, the transaction is rolled back and the consumer is repositioned so that the rolled-back record(s) can be retrieved on the next poll.
+See <<after-rollback>> for more information and for handling records that repeatedly fail.
+
+Using transactions enables exactly once semantics (EOS).
+
+This means that, for a `read->process-write` sequence, it is guaranteed that the **sequence** is completed exactly once.
+(The read and process are have at least once semantics).
+
+Kafka version 2.5 now supports two EOS modes:
+
+* `ALPHA` - aka `transactional.id` fencing (since version 0.11.0.0)
+* `BETA` - aka fetch-offset-request fencing (since version 2.5)
+
+With mode `ALPHA`, the producer is "fenced" if another instance with the same `transactional.id` is started.
+Spring manages this by using a `Producer` for each `group.id/topic/partition`; when a rebalance occurs a new instance will use the same `transactional.id` and the old producer is fenced.
+
+With mode `BETA`, it is not necessary to have a producer for each `group.id/topic/partition` because consumer metadata is sent along with the offsets to the transaction and the broker can determine if the producer is fenced using that information instead.
+
+To configure the container to use mode `BETA`, set the container property `EOSMode` to `BETA`.
+
+IMPORTANT: With `BETA`, your brokers must be version 2.5 or later, unless you are using `kafka-clients` version 2.6; the producer in that version can automatically fall back to `ALPHA` if the broker does not support `BETA`.
+To enable the fall back in the 2.6 client; set `internal.auto.downgrade.txn.commit=true`.
+
+You should also set the `DefaultKafkaConsumerFactory` `producerPerConsumerPartition` property to `false`, to reduce the number of producers needed.
+
+If your brokers are upgraded to 2.5, you can immediately enable `BETA` mode.
+
+If your brokers are older than 2.5, when the 2.6 client is available, you can enable `BETA` mode. but leave the `producerPerConsumerPartition` to its default `true`.
+When your brokers are upgraded to 2.5 or later, the producer will automatically switch to using mode `BETA`, but the number of producers will remain as before.
+You can then do a rolling upgrade of your application with `producerPerConsumerPartition` set to `false` to reduce the number of producers.
```

Comment:
Don't we need then set this property to `true` by default in Spring Kafka `2.6` ?

Reply:
No; we can't - it must be set by the user when his broker is upgraded to 2.5 - there is currently no API for us to find out the broker version. However, they are working on [KIP-584](https://cwiki.apache.org/confluence/display/KAFKA/KIP-584%3A+Versioning+scheme+for+features) so it might be possible in future.

We will, however, switch the EOSMode to BETA in 2.6.